java/component/keyple-core/src/main/java/org/eclipse/keyple/core/seproxy/plugin/reader/AbstractLocalReader.java

```diff
@@ -458,323 +324,301 @@ public void setSeProtocolSetting(Map<SeProtocol, String> protocolSetting) {
-          }
-          /* close the channel (will reset the current selection status) */
-          closeLogicalChannel();
-        }
-        /* keep the current selection status (may be null if the current PO didn't match) */
-        selectionStatus = currentSelectionStatus;
-      }
 
-      /* open the channel and do the selection if needed */
-      if (!logicalChannelIsOpen) {
-        previouslyOpen = false;
+    if (fciResponse.isSuccessful() && fciResponse.getDataOut().length == 0) {
+      /*
+       * The selection didn't provide data (e.g. OMAPI), we get the FCI using a Get Data
+       * command.
+       *
+       * The AID selector is provided to handle successful status word in the Get Data
+       * command.
+       */
+      fciResponse = recoverSelectionFciData(aidSelector);
+    }
+    return fciResponse;
+  }
 
-        try {
-          selectionStatus = openLogicalChannelAndSelect(seRequest.getSeSelector());
-          if (logger.isTraceEnabled()) {
-            logger.trace(
-                "[{}] processSeRequest => Logical channel opening success.", this.getName());
-          }
-        } catch (IllegalArgumentException e) {
-          if (logger.isDebugEnabled()) {
-            logger.debug(
-                "[{}] processSeRequest => Logical channel opening failure", this.getName());
-          }
-          closeLogicalChannel();
-          /* return a null SeResponse when the opening of the logical channel failed */
-          return null;
-        }
+  /**
+   * Select the SE according to the {@link SeSelector}.
+   *
+   * <p>The selection status is returned.<br>
+   * 3 levels of filtering/selection are applied successively if they are enabled: protocol, ATR and
+   * AID.<br>
+   * As soon as one of these operations fails, the method returns with a failed selection status.
+   *
+   * @param seSelector A not null {@link SeSelector}.
+   * @return A not null {@link SelectionStatus}.
+   * @see #processSeRequestLogical(SeRequest)
+   */
+  private SelectionStatus processSelection(SeSelector seSelector) {
 
-        if (selectionStatus.hasMatched()) {
-          /* The selection process succeeded, the logical channel is open */
-          logicalChannelIsOpen = true;
+    AnswerToReset answerToReset;
+    ApduResponse fciResponse;
+    boolean hasMatched = true;
 
-          if (selectionStatus.getFci().isSuccessful()) {
-            /* the selection AID based was successful, keep the aid */
-            aidCurrentlySelected = seRequest.getSeSelector().getAidSelector().getAidToSelect();
-          }
-          currentSelectionStatus = selectionStatus;
+    // check protocol if enabled
+    if (seSelector.getSeProtocol() != null && !protocolFlagMatches(seSelector.getSeProtocol())) {
+      answerToReset = null;
+      fciResponse = null;
+      hasMatched = false;
+    } else {
+      // protocol check succeeded, check ATR if enabled
+      SeSelector.AtrFilter atrFilter = seSelector.getAtrFilter();
```

Comment:
We should check:

  if (seSelector.getAtrFilter() != null

Reply:
fixed. Was done in checkAtr.